src/data/tutorials.rs

```diff
@@ -0,0 +1,205 @@
+}
+
+impl TutorialLessonChain {
+    pub fn add(db: &mut Db, key: Token, block: Block) {
+        let gamestate_tutorial =
+            block.get_field_bool("save_progress_in_gamestate").unwrap_or(false);
+        db.add(Item::TutorialLessonChain, key, block, Box::new(Self { gamestate_tutorial }));
+    }
+}
+
+impl DbKind for TutorialLessonChain {
+    fn validate(&self, _key: &Token, block: &Block, data: &Everything) {
+        let mut vd = Validator::new(block, data);
+
+        vd.field_validated_key_block("trigger", |key, block, data| {
+            // TODO: verify root scope
+            // All the examples in vanilla work with Scopes::None
+            let mut sc = ScopeContext::new(Scopes::None, key);
+            validate_trigger(block, data, &mut sc, Tooltipped::No);
+        });
+
+        vd.field_integer("delay");
+        vd.field_bool("save_progress_in_gamestate");
+    }
+
+    fn has_property(
+        &self,
+        _key: &Token,
+        _block: &Block,
+        property: &str,
+        _data: &Everything,
+    ) -> bool {
+        property == "gamestate_tutorial" && self.gamestate_tutorial
+    }
+}
+
+/// These are added by the [`TutorialLesson`] item loader
+#[derive(Clone, Debug)]
+pub struct TutorialLessonStep {
+    chain: Option<Token>,
+}
+
+impl DbKind for TutorialLessonStep {
+    fn validate(&self, key: &Token, block: &Block, data: &Everything) {
+        let mut vd = Validator::new(block, data);
+
+        data.verify_exists(Item::Localization, key);
+
+        vd.field_item("text", Item::Localization);
+        vd.field_item("header_info", Item::Localization); // undocumented
+                                                          // see gui_tag in TutorialLesson
+        vd.field_value("gui_tag");
+        vd.field_value("window_name");
+        // TODO: check that this is a widget name
+        vd.multi_field_value("highlight_widget");
+
+        // TODO: These two are not used in vanilla and the docs are a bit unclear
+        vd.field_item("soundeffect", Item::Sound);
+        vd.field_item("voice", Item::Sound);
+
+        vd.field_bool("repeat_sound_effect");
+        vd.field_integer("delay");
+        vd.field_value("animation");
+        vd.field_bool("shown_in_encyclopedia");
+
+        vd.multi_field_validated_block("gui_transition", |block, data| {
+            let mut vd = Validator::new(block, data);
+            vd.field_value("button_id");
+            vd.field_item("button_text", Item::Localization);
+            vd.field_validated_value("target", validate_lesson_target);
+            vd.field_validated_key_block("enabled", |key, block, data| {
+                let mut sc = ScopeContext::new(game_tutorial_scope(), key);
+                validate_trigger(block, data, &mut sc, Tooltipped::No);
+            });
+        });
+        vd.multi_field_validated_block("trigger_transition", validate_trigger_transition);
+
+        vd.field_validated_key_block("interface_effect", |key, block, data| {
```

Comment:
not present in vic3 doc... best to verify

Reply:
I added some TODO :)